File: Top_100_Liked_Questions/437._Path_Sum_III/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number}
 */
var pathSum = function(root, sum) {
    if (!root) {
        return 0;
    }
    
    let ans = 0;
    const numPath = (r, s=sum) => {
        if (!r) {
            return
        }
        
        if (r.val === s){
            ans += 1;
        }
        
        numPath(r.left, s - r.val);
        numPath(r.right, s - r.val);
    }
    
//     let currNode = root;
//     const stack = [root];
    
//     while (stack.length) {
//         if (currNode) {
            
//             if (currNode.right) {
//                 stack.push(currNode.right);
//             }
            
//             numPath(currNode);
//             currNode = currNode.left;
//         } else {
//             currNode = stack.pop();
//         }
//     }
    
    const queue = [root];
    while (queue.length) {
        const currNode = queue.shift();
        
        numPath(currNode);
        
        if (currNode.left) {
            queue.push(currNode.left);
        }
        
        if (currNode.right) {
            queue.push(currNode.right);
        }
    }
    
    return ans;
};

File: Top_100_Liked_Questions/437._Path_Sum_III/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        if not root:
            return 0
        
        ans = [0]
        
        # count the paths with the starting point at the root
        def helper(root, s=sum):
            if not root:
                return
            
            if root.val == s:
                ans[0] += 1
            
            helper(root.left, s - root.val)
            helper(root.right, s - root.val)
        
        # traverse the tree and call the above function for every node
        
        # preorder traversal
#         currNode = root
#         stack = [root]
#         while stack:
#             if currNode:
#                 if currNode.right:
#                     stack.append(currNode.right)
            
#                 helper(currNode)
#                 currNode = currNode.left
            
#             else:
#                 currNode = stack.pop()
        
        # level-order traversal
        queue = [root]
        while queue:
            currNode = queue.pop(0)
            helper(currNode)
            
            if currNode.left:
                queue.append(currNode.left)

            if currNode.right:
                queue.append(currNode.right)
            
        return ans[0]